test/lit/passes/gufa-vs-cfp.wast

```diff
@@ -0,0 +1,2753 @@
+;; name getting in the way)
+
+;; This is almost identical to cfp.wast, and is meant to facilitate comparisons
+;; between the passes - in particular, gufa should do everything cfp can do,
+;; although it may do it differently. Changes include:
+;;
+;;  * Tests must avoid things gufa optimizes away that would make the test
+;;    irrelevant. In particular, parameters to functions that are never called
+;;    will be turned to unreachable by gufa, so instead make those calls to
+;;    imports. Gufa will also realize that passing ref.null as the reference of
+;;    a struct.get/set will trap, so we must actually allocate something.
+;;  * Gufa optimizes in a more general way. Cfp will turn a struct.get whose
+;;    value it infers into a ref.as_non_null (to preserve the trap if the ref is
+;;    null) followed by the constant. Gufa has no special handling for
+;;    struct.get, so it will use its normal pattern there, of a drop of the
+;;    struct.get followed by the constant. (Other passes can remove the
+;;    dropped operation, like vacuum in trapsNeverHappen mode).
+;;  * Gufa's more general optimizations can remove more unreachable code, as it
+;;    checks for effects (and removes effectless code).
+;;
+;; This file could also run cfp in addition to gufa, but the aforementioned
+;; changes cause cfp to behave differently in some cases, which could lead to
+;; more confusion than benefit - the reader would not be able to compare the two
+;; outputs and see cfp as "correct" which gufa should match.
+;;
+;; Note that there is some overlap with gufa-refs.wast in some places, but
+;; intentionally no tests are removed here compared to cfp.wast, to make it
+;; simple to map the original cfp tests to their ported versions here.
+
+(module
+  (type $struct (struct i32))
+  ;; CHECK:      (type $none_=>_none (func_subtype func))
+
+  ;; CHECK:      (func $impossible-get (type $none_=>_none)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (unreachable)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $impossible-get
+    (drop
+      ;; This type is never created, so a get is impossible, and we will trap
+      ;; anyhow. So we can turn this into an unreachable.
+      (struct.get $struct 0
+        (ref.null $struct)
+      )
+    )
+  )
+)
+
+(module
+  (type $struct (struct i64))
+  ;; CHECK:      (type $none_=>_none (func_subtype func))
+
+  ;; CHECK:      (func $test (type $none_=>_none)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i64.const 0)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $test
+    ;; The only place this type is created is with a default value, and so we
+    ;; can optimize the get into a constant (note that no drop of the
+    ;; ref is needed: the optimizer can see that the struct.get cannot trap, as
+    ;; its reference is non-nullable).
+    (drop
+      (struct.get $struct 0
+        (struct.new_default_with_rtt $struct
+          (rtt.canon $struct)
+        )
+      )
+    )
+  )
```

Comment:
The specific reason in the test you mentioned is that dropping each result leads to GUFA seeing that the result does not reach anywhere. And that null is an opaque placeholder for CFP, but GUFA sees that it will trap. So we need to connect results to uses, which "compresses" the test.

More details here:

https://github.com/WebAssembly/binaryen/blob/adef90c1e53a163381537939f9640228221f8a43/test/lit/passes/gufa-vs-cfp.wast#L6-L22